docs/contribute/best-practices.md

```diff
@@ -0,0 +1,742 @@
+        incorrectly they cannot honor the contract anymore). Use the
+        following to tell callers that you do not honor the contract:\
+```java
+            /**
+             * @nonInterruptable This method does not support being interrupted
+             */
+```
+-   BTW: There is no obligation in the contract for you to exit as
+    quickly as possible.
+-   For more read:
+    <http://www-128.ibm.com/developerworks/java/library/j-jtp05236.md>
+
+## Broadcast Listener
+
+To avoid repeated code blocks in an Observable like
+```java
+class Observable {
+ 
+  List listeners = new ...
+  public void addListener(Listener listener){listeners.add(listener)}
+  public void removeListener(Listener listener){...}
+ 
+  public void someMethod() {
+    ...
+    // notify all listeners
+    for (Listener l : listeners) {
+      l.notify();
+    }
+  }
+ 
+  public void someMethod2() {
+    ...
+    // notify all listeners again
+    for (Listener l : listeners) {
+     l.notify();
+    }
+  }
+ 
+}
+```
+It is recommended to use a helper class `BroadcastListener` that
+provides a method to notify all its registered listeners. The
+`BroadcastListener` should be a singleton managed by `PicoContainer`.
+```java
+public class BroadcastListener implements Listener {
+ 
+  List listeners = new ...
+  public void add(Listener listener){listeners.add(listener)}
+  public void remove(Listener listener){...}
+ 
+  public void notify() {
+    for (Listener l : listeners) {
+      l.notify();
+    }
+  }
+ 
+}
+```
+
+The code for an Observable becomes therfore much simpler. It only needs
+to know the `BroadcastListener` and can easily notify all registered
+listeners at once.
+```java
+class Observable {
+ 
+  BroadcastListener listener = new BroadcastListener();
+ 
+  public void someMethod(){
+    ...
+    listener.notify();
+  }
+ 
+  public void someMethod2(){
+    ...
+    listener.notify();
+   }
+}
+```
+
+
+## Software Design Rules
+
+-   Distinguish clearly between the common building blocks of
+    applications, namely:
+    -   Services - Providers of functionality that can be well
+        encapsulated
+    -   Entities - Mutable classes that represent the business world.
+        Group these into aggregates.
+    -   Values - Use the [Value Object
+        Pattern](http://www.c2.com/cgi/wiki?ValueObject) to ensure that
+        users of the class can rely on the objects to be immutable and
+        all methods to be side effect free. This helps A LOT when using
+        a class.
+-   Avoid implementing many interfaces in the same class whereever
+    possible (more than one is necessary only in rare cases). Use nested
+    or anonymous classes instead. It makes it much easier to understand
+    and modify your code.
+
+<!-- -->
+
+-   The Singleton Pattern is a inferior solution to define the
+    architecture of an application compared to using Dependency
+    Injection (see next section). To achieve the lazy semantics often
```

Comment:
This text should link to a technical documentation explaining our internal dependency management with PicoContainer. This documentation is not contained in this PR, because we have to review this part together. I will remove this text for now.